airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalClient.java

```diff
@@ -469,23 +454,61 @@ private <T> T getWorkflowOptionsWithWorkflowId(final Class<T> workflowClass, fin
     return client.newWorkflowStub(workflowClass, TemporalUtils.getWorkflowOptionsWithWorkflowId(jobType, name));
   }
 
-  private <T> T getExistingWorkflow(final Class<T> workflowClass, final String name) {
-    return client.newWorkflowStub(workflowClass, name);
-  }
-
-  ConnectionManagerWorkflow getConnectionUpdateWorkflow(final UUID connectionId) {
-    final boolean workflowReachable = isWorkflowReachable(getConnectionManagerName(connectionId));
-
-    if (!workflowReachable) {
-      throw new IllegalStateException("No reachable workflow for the connection {} while trying to delete it");
+  /**
+   * Attempts to retrieve the connection manager workflow for the provided connection.
+   *
+   * @param connectionId the ID of the connection whose workflow should be retrieved
+   * @return the healthy ConnectionManagerWorkflow
+   * @throws DeletedWorkflowException if the workflow was deleted, according to the workflow state
+   * @throws UnreachableWorkflowException if the workflow is unreachable
+   */
+  private ConnectionManagerWorkflow getConnectionManagerWorkflow(final UUID connectionId)
+      throws DeletedWorkflowException, UnreachableWorkflowException {
+    final ConnectionManagerWorkflow connectionManagerWorkflow;
+    final WorkflowState workflowState;
+    try {
+      connectionManagerWorkflow = client.newWorkflowStub(ConnectionManagerWorkflow.class, getConnectionManagerName(connectionId));
+      workflowState = connectionManagerWorkflow.getState();
+    } catch (final Exception e) {
+      throw new UnreachableWorkflowException(
+          String.format("Failed to retrieve ConnectionManagerWorkflow for connection %s due to the following error:", connectionId),
+          e);
     }
 
-    final ConnectionManagerWorkflow connectionManagerWorkflow =
-        getExistingWorkflow(ConnectionManagerWorkflow.class, getConnectionManagerName(connectionId));
+    if (workflowState.isDeleted()) {
```

Comment:
I don't understand how this block is working. I means than for most of the deleted workflow, we will consider them as Unreachable because the workflow are suppose to be terminated.

Reply:
@benmoriceau One thing I discovered while investigating this issue is that when we delete a connection, the temporal workflow is actually **not** terminated. The temporal workflow in the delete case has status `Completed`, which is a separate status from `Terminated`. For these deleted workflows, we can actually still retrieve the workflow and ready values from its workflowState, but we cannot call any signal methods on the workflow because it is not actively running.

I think the reason the temporal workflow is `Completed` and not `Terminated` in the deletion case is because we just `return` in that case: https://github.com/airbytehq/airbyte/blob/f816946a09a890c553ee42550d86d6b6ed4b231a/airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java#L105-L109
And I think when the `run` method `return`s, temporal marks the workflow as `Completed`

Reply:
So my new understanding is that the "unreachable" case is not ever really expected, which is why we repair the workflow in that case

Reply:
I see and for a `Terminated` workflow, I guess we can't query the state.

Reply:
Yep that's correct

Reply:
Thank, approved then.